pilot/pkg/model/virtualservice.go

```diff
@@ -0,0 +1,360 @@
+// Copyright 2020 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"strings"
+
+	"github.com/gogo/protobuf/jsonpb"
+
+	networking "istio.io/api/networking/v1alpha3"
+
+	"istio.io/istio/pilot/pkg/features"
+	"istio.io/istio/pilot/pkg/util/sets"
+)
+
+func mergeVirtualServicesIfNeeded(vServices []Config) (out []Config) {
+	out = make([]Config, 0, len(vServices))
+	delegatesMap := map[string]Config{}
+	// root virtualservices with delegate
+	var rootVses []Config
+
+	// 1. classify virtualservices
+	for _, vs := range vServices {
+		rule := vs.Spec.(*networking.VirtualService)
+		// it is delegate, add it to the indexer cache
+		if len(rule.Hosts) == 0 {
+			delegatesMap[key(vs.Name, vs.Namespace)] = vs
+			continue
+		}
+
+		// root vs
+		if isRootVs(rule) {
+			rootVses = append(rootVses, vs)
+			continue
+		}
+
+		// the others are normal vs without delegate
+		out = append(out, vs)
+	}
+
+	// If `PILOT_ENABLE_VIRTUAL_SERVICE_DELEGATE` feature disabled,
+	// filter out invalid vs(root or delegate), this can happen after enable -> disable
+	if !features.EnableVirtualServiceDelegate {
+		return
+	}
+
+	// 2. merge delegates and root
+	for _, root := range rootVses {
+		rootVs := root.Spec.(*networking.VirtualService)
+		mergedRoutes := []*networking.HTTPRoute{}
+		for _, route := range rootVs.Http {
+			// it is root vs with delegate
+			if route.Delegate != nil {
+				delegate, ok := delegatesMap[key(route.Delegate.Name, route.Delegate.Namespace)]
+				if !ok {
+					log.Debugf("delegate virtual service %s/%s of %s/%s not found",
+						route.Delegate.Namespace, route.Delegate.Name, root.Namespace, root.Name)
+					// delegate not found, ignore only the current HTTP route
+					continue
+				}
+				// DeepCopy to prevent mutate the original delegate, it can conflict
+				// when multiple routes delegate to one single VS.
+				copiedDelegate := delegate.DeepCopy()
```

Comment:
are you sure we need this copy? we already do a copy in https://github.com/istio/istio/pull/22118/files#diff-ee39cd332931ab6f8c10713ce3446d09R1130

Reply:
ignore